kafka/main.go

```diff
@@ -67,92 +88,145 @@ func StartConsumers(providedServer *server.Server, logger *zerolog.Logger) error
+// processMessagesIntoBatchPipeline fetches messages from Kafka indefinitely, pushes a
+// MessageContext into the batchPipeline to maintain message order, and then spawns a
+// goroutine that will process the message and push to errorResult of the MessageContext
+// when the processing completes. There *must* be a value pushed to the errorResult, so
+// a simple ProcessingError is created for the success case.
+func processMessagesIntoBatchPipeline(
+	ctx context.Context,
+	topicMappings []TopicMapping,
+	providedServer *server.Server,
+	reader *kafka.Reader,
+	batchPipeline chan *MessageContext,
+	logger *zerolog.Logger,
+) {
+	// During normal operation processMessagesIntoBatchPipeline will never complete and
+	// this deferral should not run. It's only called if we encounter some unrecoverable
+	// error.
+	defer func() {
+		close(batchPipeline)
+	}()
 
-	for i := 1; i <= consumerCount; i++ {
-		go func(topicMappings []TopicMapping) {
-			consumer := newConsumer(topics, adsConsumerGroupV1, logger)
-			var (
-				failureCount = 0
-				failureLimit = 10
-			)
-			logger.Trace().Msg("Beginning message processing")
-			for {
-				// `FetchMessage` blocks until the next event. Do not block main.
-				ctx := context.Background()
-				logger.Trace().Msgf("Fetching messages from Kafka")
-				msg, err := consumer.FetchMessage(ctx)
-				if err != nil {
-					logger.Error().Err(err).Msg("")
-					if failureCount > failureLimit {
-						break
-					}
-					failureCount++
-					continue
-				}
-				logger.Info().Msgf("Processing message for topic %s at offset %d", msg.Topic, msg.Offset)
-				logger.Info().Msgf("Reader Stats: %#v", consumer.Stats())
-				for _, topicMapping := range topicMappings {
-					if msg.Topic == topicMapping.Topic {
-						go func(
-							msg kafka.Message,
-							topicMapping TopicMapping,
-							providedServer *server.Server,
-							logger *zerolog.Logger,
-						) {
-							err := topicMapping.Processor(
-								msg.Value,
-								topicMapping.ResultProducer,
-								providedServer,
-								logger,
-							)
-							if err != nil {
-								logger.Error().Err(err).Msg("Processing failed.")
-							}
-						}(msg, topicMapping, providedServer, logger)
-
-						if err := consumer.CommitMessages(ctx, msg); err != nil {
-							logger.Error().Msgf("Failed to commit: %s", err)
-						}
-					}
-				}
+	for {
+		msg, err := reader.FetchMessage(ctx)
+		if err != nil {
+			// Indicates batch has no more messages. End the loop for
+			// this batch and fetch another.
+			if err == io.EOF {
+				logger.Info().Msg("Batch complete")
+			} else if strings.ToLower(err.Error()) != "context deadline exceeded" {
+				logger.Error().Err(err).Msg("batch item error")
+				// @TODO: Is there a way to close the batchPipeline and
+				// allow the MessageContexts in it to complete before we
+				// panic? Panic here will only stop this goroutine.
+				panic("failed to fetch kafka messages and closed channel")
```

Comment:
seems like we can just call `close(batchPipeline)` and use the two value form of receive as shown here - https://gobyexample.com/closing-channels

Reply:
do we still need to use the two value form of recieve?